### DISCUSSION TITLE
Accessing Request body as bytes in custom middleware

### DISCUSSION THREAD

**Discussion by theelderbeever** [tokio-rs/axum#2265] [Q&A] [answered]
I am writing a custom middleware for verifying an hmac signature header by reproducing the signature from a secret and the body of the request being sent. Below is the code I am using and would like to pass to `.layer(middleware::from_fn(...))` in my router.

The issue I am running into is that I need to access the body of the request as bytes or a String however, the generic `<B>` that seems to be required by the function signature doesn't have a way to convert to any other types. I am sure its just a misunderstanding of how all this works but, how can I go about actually accessing the body of the `Request` as some actually usable type?

```rust
async fn verify_orb_signature<B>(
    headers: HeaderMap,
    request: Request<B>,
    next: Next<B>,
) -> Result<Response, StatusCode> {
    let (orb_timestamp, orb_signature) = match (
        headers.get("x-orb-timestamp"),
        headers.get("x-orb-signature"),
    ) {
        (Some(ts), Some(sig)) => (ts.to_str().unwrap(), sig.as_bytes()),
        _ => return Err(StatusCode::BAD_REQUEST),
    };

    let secret = std::env::var("ORB_WEBHOOK_SECRET").unwrap();

    // let body = request.body();

    // let verified = webhooks::verify_signature(orb_signature, orb_timestamp, secret, body);
    // if !verified {
    //     return Err(StatusCode::BAD_REQUEST);
    // }
    let response = next.run(request).await;
    Ok(response)
}
```

**Comment by davidpdrsn** [comment, marked as answer]
Have a look at https://github.com/tokio-rs/axum/blob/axum-v0.6.20/examples/consume-body-in-extractor-or-middleware/src/main.rs

**Reply by theelderbeever** [reply]
Awesome! Thank you!
